File: 6 kyu/PI approximation/solution.cs
using System.Collections;

using static System.Math;

public class PiApprox
{
    public static ArrayList iterPi(double epsilon)
    {
        var sum = 0.0;
        var i = 0;

        do
        {
            sum += (i % 2 == 0 ? 4.0 : -4.0) / (2 * i + 1);
            ++i;
        }
        while (Abs(PI - sum) >= epsilon);

        return new ArrayList { i, Round(sum, 10) };
    }
}
________________________________________
using System;
using System.Collections;

public static class PiApprox
{
  public static ArrayList iterPi(double epsilon)
  {
    double myPi = 0.0;
    int i = 0;
    while (Math.Abs(myPi - Math.PI) > epsilon)
      myPi += 4 * Math.Pow(-1, i) / (2 * i++ + 1);
    return new ArrayList {i, Math.Round(myPi, 10, MidpointRounding.AwayFromZero)};
  }
}
________________________________________
using System;
using System.Collections;
public class PiApprox
        {
            public static ArrayList iterPi(double epsilon)
            {
                 double myPI = 0d;
            double sum = 0d;
            int n = 0;
            while (Math.Abs(Math.PI - myPI) > epsilon)
            {
                sum = sum + Math.Pow(-1,n) / (2 * n + 1);
                n++;
                myPI = 4 * sum;
            }
           
            return new ArrayList{n, Math.Round(myPI,10)};
            }
        }
________________________________________
using System.Collections;
using System;
public class PiApprox
        {
            public static ArrayList iterPi(double epsilon)
            {
                double pp = 0 , ci = (1/epsilon>=100000 ? 1/epsilon+1 : 1/epsilon);
                for ( double i=0 , db=1 ; i<ci ; i++ , db+=2 ) pp+=( i%2==1 ? -1/db : 1/db );
                return new ArrayList{ Math.Round(ci,0) , Math.Round(pp*4,10) };
            }
        }
________________________________________
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;

public class PiApprox
        {
            public static ArrayList iterPi(double epsilon)
            {
                ArrayList res = new ArrayList { };
                double n = 1.0; double value = 0; int counter = 0;
                while (Math.Abs((Math.PI - 4 * value)) > epsilon)
                {
                    value += 1.0 / n;
                    n = -n;
                    if (n > 0) n += 2.0;
                    if (n < 0) n -= 2;
                    counter += 1;
                }
                res.Add(counter); res.Add(Math.Round(4 * value, 10));
                return res;
            }
        }

File: 6 kyu/PI approximation/solution.clj
(ns piapprox.core)

(defn round [s n] 
  (.setScale (bigdec n) s java.math.RoundingMode/HALF_EVEN)) 

(defn iterPi [epsilon]
  (loop [i 1 value 4.0]
    (if (<= (Math/abs (- Math/PI value)) epsilon)
      [i, (round 10 value)]
      (do
        ;(println value)
        (recur (inc i) (+ value (* 4.0 (/ (Math/pow -1 i) (+ 1 (* 2 i))))))))))
________________________________________
(ns piapprox.core)

(defn round [s n] 
  (.setScale (bigdec n) s java.math.RoundingMode/HALF_EVEN)) 

(defn iterPi [epsilon]
  (loop [i 1 pi4 1.0 s -1.0 d 3.0]
    (if (<= (Math/abs (- (* 4 pi4) Math/PI)) epsilon)
      [i, (round 10 (* 4 pi4))]
      (recur (inc i) (+ pi4 (* s (/ 1.0 d))) (* s -1.0) (+ 2 d)))))
________________________________________
(ns piapprox.core)

(defn round [s n]
  (.setScale (bigdec n) s java.math.RoundingMode/HALF_EVEN))

(defn iterPi [epsilon]
  (let [too-inaccurate (fn [[_ x]] (> (Math/abs (- x Math/PI)) epsilon))
        series-element (fn [k] (/ (* (- 1 (* 2 (mod k 2))) 4.0) (+ (* 2 k) 1)))
        [[n s]] (take 1 (drop-while too-inaccurate
                              (reductions (fn [[_ s] [k x]] [k (+ s x)])
                                          (map (fn [k] [k (series-element k)]) (range)))))]
    [(inc n) (round 10 s)]))
________________________________________
(ns piapprox.core)

(defn round [s n] 
  (.setScale (bigdec n) s java.math.RoundingMode/HALF_EVEN)) 

(defn iterPi [epsilon]
  (loop [acc 4.0
         sgn -1
         div 3
         num 1]
    (if (< (Math/abs (- Math/PI acc)) epsilon)
      [num (round 10 acc)]
      (recur (+ acc (/ (* 4 sgn) div))
             (- sgn)
             (+ div 2)
             (inc num)))))
________________________________________
(ns piapprox.core)

(defn round [s n] 
  (.setScale (bigdec n) s java.math.RoundingMode/HALF_EVEN)) 

(defn iterPi [epsilon]
  (->> (iterate 
          #(let [[val num sig]  %
                 den            (inc (* num 2))
                 add            (* sig (/ 1 den))]
              [(+ add val) (inc num) (* sig -1)])
              [1.0 1 -1])
       (drop-while #(> (Math/abs (- (* 4 (first %)) Math/PI)) epsilon))
       (first)
       ((fn [[val num _]] [num (round 10 (* 4 val))]))
  )
)
